File: gomoku.py
#!/bin/python

from math import *
from time import time
from print_grid import *
from win_condition import print_winner
from minimax_ab import calculateNextMove
from best_move import my_print

ia_move = []

def all_print(arr, turn, tmp):
    global ia_move
    print_grid(arr)
    print_winner(arr)
    if turn % 2 == 0 and turn > 0:
        print("IA played ", end="")
        my_print(ia_move)
    if (tmp == turn):
        print("Error retry")
    if turn % 2 == 0:
        print("Player 1\nTurn:", turn + 1)
    else:
        print("Player 2\nTurn:", turn + 1)
        print("IA is thinking...")
    return turn

def verify_intput(ltr, nb):
    if ltr < 0 or ltr > 18:
        return 1
    if nb < 0 or nb > 18:
        return 1
    return 0

def ai_play(grid):
    global ia_move

    move = calculateNextMove(grid, 4)
    grid[move[0]][move[1]] = -1

    ia_move = move
    return grid

def main():
    arr = [[0 for i in range(19)] for j in range(19)]
    get = []
    turn = 0
    tmp = -1

    while True:
        tmp = all_print(arr, turn, tmp)
        if turn % 2:
            arr = ai_play(arr)
            turn += 1
            continue
        get = (str(input()))
        if len(get) > 3:
            continue
        try:
            ltr = ord(get[0][0]) - 65
            nb = int(get[1:]) - 1
        except:
            continue
        if ltr > 18:
            ltr -= 32
        if (verify_intput(ltr, nb) == 1):
            continue
        if arr[nb][ltr] == 0:
            if turn % 2 == 0:
                arr[nb][ltr] = 1
            else:
                arr[nb][ltr] = -1
        else:
            continue
        turn += 1

main()

File: cpp/gomoku.cpp
#include "gomoku.hpp"
#include <stdio.h>
#include <string.h>
#include <cstdlib>
#include <utility>
#include <vector>
#include <iostream>
#include <string>

move_t ia_move;

void ai_play(char grid[SIZE][SIZE])
{
    move_t move = calculateNextMove(grid, 3);

    grid[move.first][move.second] = -1;
    ia_move = move;
}

void print_ia_move(void)
{
    std::cout << "IA played " <<
        char('A' + ia_move.first) << ia_move.second + 1
    << std::endl;
}

int all_print(char grid[SIZE][SIZE], int turn, int tmp)
{
    print_grid(grid);
    print_winner(grid);
    if (turn % 2 == 0 && turn > 0)
        print_ia_move();
    if (tmp == turn)
        printf("Error retry\n");
    if (turn % 2 == 0)
        printf("Player 1\nTurn:%i\n", turn + 1);
    else {
        printf("Player 2\nTurn:%i\n", turn + 1);
        printf("IA is thinking...\n");
    }
    return turn;
}

bool verify_intput(int &ltr, int nb)
{
    if (ltr > SIZE - 1)
        ltr -= 32;
    if (ltr < 0 or ltr > SIZE - 1)
        return true;
    if (nb < 0 or nb > SIZE - 1)
        return true;
    return false;
}

int main(void)
{
    char grid[SIZE][SIZE];
    char *get = nullptr;
    size_t s = 0;
    int turn = 0;
    int tmp = -1;

    for (int i = 0; i < SIZE; i++)
        memset(grid[i], 0, sizeof(char) * SIZE);
    for (;;) {
        tmp = all_print(grid, turn, tmp);
        if (turn % 2) {
            ai_play(grid);
            turn += 1;
            continue;
        }
        free(get);
        get = nullptr;
        if (getline(&get, &s, stdin) < 0)
            return 0;
        if (!get || strlen(get) > 4 || strlen(get) < 3)
            continue;
        int ltr = get[0] - 65;
        int nb = atoi(get + 1) - 1;
        if (verify_intput(ltr, nb))
            continue;
        if (grid[nb][ltr] == 0) {
            if (turn % 2 == 0)
                grid[nb][ltr] = 1;
            else
                grid[nb][ltr] = -1;
        } else
            continue;
        turn += 1;
    }
    return 0;
}
